Random Ray Kinetic Simulation Mode

yardasol opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

This PR adds a kinetic simulation mode to OpenMC's random ray solver to allow simulation of kinetic transients (i.e. time-dependent transients without temperature/reactivity feedback)

Specifically, this PR adds the following:
- A new `kinetic_simulation` toggle setting.
- Material density transient capabilities.
- ~A new multi-group cross section type, `CHI`, to complement `CHI_PROMPT` and `CHI_DELAYED`~ Superceded by #3740 
- A new score type, `PRECURSORS`
- ~A new example model, `random_ray_pin_cell`~ Superceded by #3735 
- ~Parameterizes the main random ray function to adhere to D.R.Y.~ Superceded by #3733 
- Relevant content in the Theory Manual
- Added kinetic random ray metrics to the statepoint file 
- New testing harnesses for testing kinetic simulations
- Automagic setup extensions to support kinetic simulations
- Regression and unit tests

I realize this is quite a large feature and may take a while to review. I'd appreciate any discussion on necessary theory/user guide additions to the doc pages. Most of what can be done with the base random ray solver can be done with the kinetic simulation mode (with the exception of simulating void regions, for now), so I figured not much past basic explanation of the feature and how to use it was needed.

# Verification
I utilized two models to verify the kinetic simulation machinery

1. A pin-cell problem I compared against point kinetics
2. The C5G7-TD Exercise 3-1 Benchmark Problem

| Pincell | C5G7 Detail|
|-----------|------------------|
|<img width="1000" height="1000" alt="pincell_cellwise" src="https://github.com/user-attachments/assets/45a0f2f9-21ba-49f1-b5fb-ccd544820e55" />|<img width="1000" height="1000" alt="c5g7-rr-cells-subregion" src="https://github.com/user-attachments/assets/0bfa158b-521e-444c-8c00-22eb21465b22" />|


The pin cell model served as a quick sanity check for the reasonability of the my implementation’s solution. A fuel pin cell is easy to model and does not take a prohibitively long time to run. It is also relatively simple to model a pin cell using point-reactor kinetics, which serve as a robust check of my approach.

The more expensive C5G7 model was chosen due to
- Being an older OECD/NEA benchmark, cross-section data and model specification is openly available. 
- An [existing C5G7 model for OpenMC](https://github.com/mit-crpg/benchmarks/tree/master/c5g7/openmc) is available.
- Many time-dependent neutron transport publications utilize the C5G7-TD suite of benchmarks to do code-to-code comparison, so there are a plethora of reference results to compare against.                       

I utilize the same transient for both the pin cell model and the C5G7 model.  The transient is based on Exercise 3-1 of the C5G7-TD benchmark. Moderator density is assumed to be at its nominal values at  t=0 s . The moderator experiences a linear density decrease until it reaches a minimal value, $\omega=0.95$ for Exercise 3-1, at  $t=1 s$ . The moderator then experiences a linear density increase until it returns to its nominal value at  $t = 2s$ . The C5G7-TD reference results continue the simulation for another 8 seconds with the moderator at its nominal value. The moderator density changes imposed in the Exercise 3-1 transient, as well as the other three sub-cases for Exercise 3, are shown in the figure below:

<img width="320" height="240" alt="c5g7-exercise3-moderator" src="https://github.com/user-attachments/assets/034e2e01-44b1-4723-bbd0-c61211b337b2" />

## Pincell Results
I compared the kinetic Random Ray implementation against [PyRK](https://github.com/pyrk/pyrk), an open-source point-kinetics code. To generate the proper feedback coefficient, I ran several k-eigenvalue simulations along the range of densities in the Exercise 3-1 transient. Below are the PyRK and OpenMC results plotted against each other:

| Flux                            | Precursors                           |
| ----------------------------------- | ----------------------------------- |
| <img width="640" height="480" alt="rr-vs-pyrk-flux" src="https://github.com/user-attachments/assets/4a740eee-1ccb-4da8-bd06-d37d0d08b183" /> | <img width="640" height="480" alt="rr-vs-pyrk-precursors" src="https://github.com/user-attachments/assets/8245a076-0027-4fdf-9228-7a0a1e39dccc" /> |

There is a large deviation between the time-dependent Random Ray solution and the point kinetics solution. While the two are very relatively close for the first leg of the density transient, once the density begins to increase the scalar flux and precursor solutions diverge. We know that the radial flux solution for a cylindrical reactor is the zeroth-order Bessel function $J_0$, which decreases in magnitude as the radial point of interest moves away from the center of the reactor. The point kinetics flattens out this spatial dependence to a singular value, so it is expected that point kinetics overestimates both the scalar flux and the precursor concentration, as seen in the above figures. Similar statements can be made about the precursor solution.

I also wanted to compare the two methods (Time Isotropic (TI) and Source Derivative Propagation (SDP)). These two methods should be equivalent for this simple test case. TI is faster but will perform poorly near highly absorbing regions. SDP makes a higher order approximation for a greater execution cost.

| Flux difference                           | Precursors  difference                         |
|--------------------------------------------|-----------------------------------------------------|
|<img width="640" height="480" alt="ti-vs-sdp-flux" src="https://github.com/user-attachments/assets/4ea04aeb-7a2f-47e8-b10c-ad6948819b34" />| <img width="640" height="480" alt="ti-vs-sdp-precursors" src="https://github.com/user-attachments/assets/3daf0e7a-53da-4484-bb86-9b4d1098b7e0" />|

 The SDP method produces a slightly lower total neutron flux at the start of the transient, on the order of $10^-3$  below the TI scalar flux , but begins to approach the TI scalar flux until the inflection point at  $t=1 s$ when it  becomes slightly larger than the SDP solution. At  $t=2 s$, the SDP solution begins approaching the TI solution again before ending slightly below it for the remainder of the simulation. These errors are small enough that it is reasonable to conclude that the SDP and TI methods are equivalent for small models.

## C5G7 results
I compared the results of the C5G7-TD transient run with OpenMC with reference data from [Shen et al. (2019)](https://doi.org/10.1016/j.anucene.2018.10.049) on the same model and transient  using the MOC code MPACT. The active length, inactive length, and rays per batch come from recommended parameters given by [Tramm](https://dspace.mit.edu/handle/1721.1/119038) for the C5G7 model. The active and inactive batches are based on numbers used by [Kraus et al. (2025)](https://doi.org/10.1080/00295639.2025.2456413). These simulation parameters are visible in the table below:

| Item | Value |
|---------|---------|                                                  
| Inactive Length [cm] | 628.0|
| Active Length [cm] | 13.0
| Batches | 10000|
| Inactive Batches | 3500|
| Rays per batch | 650|
| Time step size [s] | 0.01|
| Backwards Difference Order | 3 |

The figures below compare the MPACT reference solution to the OpenMC implementation of TI and SDP. All three methods are virtually indistinguishable during the density decrease and increase. The absolute error reaches a negative maximum of  -0.02  and a positive maximum of 0.02 at the end of the density ramp. During the flat part of the density transient ( t=2 s and onward), the normalized fission rates of the SDP and TI methods are  \sim0.05  below the MPACT normalized fission rate. Notice the oscillations in the absolute error. This is due to the stochastic nature of time-dependent Random Ray. The shape of the error plot and oscillations are very similar to Kraus’ findings.

| Fission Rate | Fission Rate Absolute Error|
|-------------------|----------------------------------------|
|<img width="640" height="480" alt="pin-ti-vs-sdp-fission-rate" src="https://github.com/user-attachments/assets/80cb91d7-5826-4bd3-8c8a-3ab60a0ec063" />|<img width="640" height="480" alt="pin-ti-vs-sdp-fission-rate-error" src="https://github.com/user-attachments/assets/3f224889-97d0-488d-994b-e5115d0c654b" />|

## Run Time Analysis
| Model | Time Derivative Method | Avg. CPU Time [s] |
|-----------|------------------------------------|---------------------------|
| Pin Cell | TI | 21.268  $\pm$  0.246 |
| Pin Cell | SDP | 22.350 $\pm$  0.234 |
| C5G7 | TI | 830.348  $\pm$  2.786
| C5G7 | SDP |1020.249 $\pm$ 2.759 |

While the TI and SDP methods are equvialent, the TI method is around $\sim 5$% and $\sim 19$% faster than the SDP method for the pin cell and C5G7 Models.

# Issues/Planned improvements
- Construction of the `tally_delay_tasks` object in `FlatSourceDomain::convert_source_regions_to_tallies` copies a lot of existing code to construct `tally_tasks`. This shared code should ideally be parameterized in a helper function. An even better solution would be generalization of `tally_tasks` to include delay groups
- There is some missing functionality, namely treatment of explicit void regions in the time-dependent case, support for time-dependent external sources, and continuing a finished simulation from a final time step. These features will be added in future PRs.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

yardasol committed:
add necessary settings and unit tests
- add `timestep_parameters` dictionary and `kinetic_simulation` switch
  to settings.py
- add `bd_order` and `time_derivative_method` to `random_ray`
  dictionary in settings.py
- add `density_timeseries` to material.py
- add corresponding variables and machinery in the C++ core
- add unit tests

yardasol committed:
preliminary machinery to process chi (total neutron energy spectrum)

yardasol committed:
timer, flat source, and source region additions

yardasol committed:
add kinetic simulation machinery
- add some missing precursor tally machinery in reaction.cpp and
  tally.cpp, and output.cpp
- add testing harness for kinetic simulations
- remove some unuesd timers
- implement time stepping loop in random_ray_simulation.cpp and
  supporting funtions.

yardasol committed:
Proper normalization for kinetic simulations

yardasol committed:
quality of life updates
- PROPOGATION -> PROPAGATION
- Write random ray data to HDF5
- Write some extra HDF5 data
- Typo fixes

yardasol committed:
make keff scaling consistent with keff normalization for kinetic simulations

Kinetic simulations based on an eigenvalue solve need to multiply the
prompt and delayed fission sources by the computed keff to bake-in the
steady state initial condition. The keff scaling introduced in PR 3595
complicates this by multiplying tallied flux by keff. This is fine,
however this new scaling factor is also applied to the final quantities
which serve as the final estimate of quantities for each timestep in
kinetic simulation. This would effetively undo the keff normalization
putting the system out of steady state. In practice this looks like a
shap step change in the values of tallied quantities. To address this,
this commit introduces control flow to ensure that the keff scaling
is applied to tallies at each timestep, but does not apply to the
internal quantitities stored in memory until the final timestep, to
ensure the keff normalization still bakes in that steady state.

yardasol committed:
move some global vars from settings.cpp to simulation.cpp
- is_initial_condition
- current_timetstep

yardasol committed:
time consistent seed approach

yardasol committed:
add kinetic simulation information to the user guide

yardasol committed:
minor formatting fix

yardasol committed:
don't store initial k_eff solve

yardasol committed:
add random_ray_k_eff_kinetic regression test

yardasol committed:
add kinetic functionality to auto setup feature

yardasol committed:
Add time-dependent mesh regression test

yardasol committed:
add blurb about kinetic automagic setup

yardasol committed:
change all source quantities back to floats

yardasol committed:
remove cruft comments

yardasol committed:
autopep8

yardasol commented:
Something is wrong with diagonal stabilization in the kinetic simulation mode. The simulation segfaults on the first k-eff simulation. Converting back to draft until I've identified and fixed the issue

yardasol committed:
fix tests, memory leak, and add more depth to userguide

yardasol commented:
It seems a code block in `FlatSourceDomain::batch_reset()` caused a memory error that made `nu_sigma_f` zero in every source element when using the auto setup workflow to generate a kinetic simulation model. Removing this block fixed this error. I'm not sure why this happens, but maybe the reason will become apparent in code review. I don't want to add undefined behavior to the main codebase so I'm looking for help on this one.

yardasol committed:
formatting fixes

nuclearkevin commented on src/random_ray/flat_source_domain.cpp:
```diff
@@ -95,12 +95,6 @@ void FlatSourceDomain::batch_reset()
     source_regions_.scalar_flux_new(se) = 0.0;
   }
-
-  if (settings::kinetic_simulation && settings::create_delayed_neutrons) {
```
I believe the issue you were running into here has to do with the indexing of `double precursors_new(int di)` (this indexes `precursors_new_` on the backend, which has length `source_regions_.n_source_regions() * ndgroups_`). `n_source_elements()` returns `source_regions_.n_source_regions() * negroups_`, which is notably quite a bit larger and so you are accidentally overrunning the precursor array.

yardasol replied:
Good catch! Not sure how I missed that, I suppose that's what I get for coding late into the night. This can be fixed by swapping out `n_source_elements()` with `n_delay_elements()`

yardasol replied:
It seems this flew under the radar previously as I was using more delay groups than energy groups for the verification. The indexing went way over when utilizing a 70-group structure, causing this odd behavior. Thanks again @nuclearkevin for your sharp eyes.

nuclearkevin replied:
:saluting_face:

yardasol committed:
fix precursor batch reset

yardasol committed:
fix failing unit test due to api change

yardasol committed:
parameterize main random ray function+
- parameterize rename_file function
- change base statepoint name used for kinetic simulations (now is
  closer to the base openmc format)
- update corresponding test infrastructure

yardasol committed:
parameterize main random ray function+
- parameterize rename_file function
- change base statepoint name used for kinetic simulations (now is
  closer to the base openmc format)
- update corresponding test infrastructure

yardasol committed:
prepare for void material support in kinetic simulations
- add void material flow control to kinetic machinery, in particular the
  calculation of source and scalar flux time derivatives
- add some missing necessary terms to propagation machinery
  (inverse_velocity multiplication)
- add SDP contributions to scalar flux in set_flux_to_flux_plus_source

yardasol committed:
use T1 variable instead of source_time_derivative and scalar_flux_time_derivative_2

yardasol committed:
use T1 variable instead of source_time_derivative and scalar_flux_time_derivative_2

yardasol committed:
add SDP void char eqs, simplify SDP scalar flux calc

jtramm commented:
Thanks @yardasol - the results look great! I'll take a look in the next week or two and will provide preliminary feedback.

yardasol committed:
pincell benchmark doesn't change significantly, so it's safe to update tests

yardasol committed:
typo fix in docs

yardasol committed:
fix sigma_t division error for sdp T1 term

yardasol committed:
adjust TI method for void regions

yardasol committed:
fix SDP

yardasol commented:
I found a small bug while moving ahead and implementing support for void regions that seems to be the cause of the SDP method's slight mismatch with the TI method. The second order scalar flux time derivative in the `T1` term must be divided by the group velocity. This seems to be missing from the current version of the code. I reran the pin cell example and found the differences between the two to decrease by half an order of magnitude. 

| Flux                            | Precursors                           |
| ----------------------------------- | ----------------------------------- |
| <img width="640" height="480" alt="rr-vs-pyrk-flux" src="https://github.com/user-attachments/assets/1d72c153-6804-4cc9-a8d4-eae59995bf5d" /> | <img width="640" height="480" alt="rr-vs-pyrk-precursors" src="https://github.com/user-attachments/assets/a120f064-670a-4ab5-9d2f-cb72bc967984" /> |

| Flux difference                           | Precursors  difference                         |
|--------------------------------------------|-----------------------------------------------------|
| <img width="640" height="480" alt="ti-vs-sdp-flux" src="https://github.com/user-attachments/assets/a7fa80bf-6180-4a26-980e-a74a77dce359" /> | <img width="640" height="480" alt="ti-vs-sdp-precursors" src="https://github.com/user-attachments/assets/09d87d9c-2b87-4428-bfe8-ae7ca60c65f9" /> |

Notice the must lower difference between the two methods compared to before. A tricky bug this was.

yardasol committed:
make source_bd type consistent with other source variables

yardasol committed:
formatting + sdp sigma_t fix

yardasol committed:
fix sigma_t division and Q storage

yardasol committed:
remove unneeded delta_psi_prime

yardasol committed:
Merge branch 'tdrr-clean-refactor' into kin-rr-sim-simplification

yardasol committed:
fix tests??

yardasol committed:
reset test results

yardasol committed:
minor fix for time_step_reset

yardasol committed:
add full kinetic test coverage

yardasol committed:
Merge branch 'tdrr-clean-refactor' into kin-rr-sim-simplification

yardasol committed:
update regression test

yardasol committed:
autopep8

yardasol committed:
autopep8

yardasol committed:
allow fixed source simulations

yardasol commented:
I added more test coverage but now those new tests I added are failing... going to run the pincell benchmark with this latest version to ensure I'm getting expected results, if they are I will update the failing tests.

yardasol committed:
update mesh kinetic tests

yardasol commented:
Pincell benchmark looked good, so I'm updating the tests.

yardasol committed:
Revert explicit void treatment in kinetic random ray

yardasol committed:
more explicitly prohibit void regions in kinetic simuations (for now)

yardasol committed:
Merge branch 'tdrr-clean-refactor' into kin-rr-sim-simplification

yardasol committed:
Merge pull request #16 from yardasol/kin-rr-sim-simplification

Consistency changes, explicitly prohibit void regions in kinetic rr until it is fully supported.

yardasol commented:
The most recent commit shows further improvement in the difference between SDP and TI because now they are truly using the same exact workflow, save for the random ray transport terms themselves.

| Flux                            | Precursors                           |
| ----------------------------------- | ----------------------------------- |
| <img width="640" height="480" alt="rr-vs-pyrk-flux" src="https://github.com/user-attachments/assets/6c0be3ab-00bd-4881-99e8-bd12289b6f11" /> | <img width="640" height="480" alt="rr-vs-pyrk-precursors" src="https://github.com/user-attachments/assets/498038c9-c52d-4191-93ee-7199cf855b62" /> |

| Flux difference                           | Precursors  difference                         |
|--------------------------------------------|-----------------------------------------------------|
| <img width="640" height="480" alt="ti-vs-sdp-flux" src="https://github.com/user-attachments/assets/4e3ad186-8789-4677-8484-652fff5f0524" /> | <img width="640" height="480" alt="ti-vs-sdp-precursors" src="https://github.com/user-attachments/assets/b5e1ea24-0763-4ec7-a700-048f768f1d8a" /> |

yardasol committed:
fix test results overwritten by merge

yardasol committed:
Merge branch 'develop' into tdrr-clean-refactor

yardasol committed:
implement density multipliers for time-dependent machinery

yardasol committed:
fix incorrect header printed

yardasol committed:
update regression test function names

yardasol commented:
For an easier review, I am splitting up the parts of this PR that are not strictly related to kinetic random ray into separate PRs. This will be a draft until these other PRs are merged.

yardasol committed:
Merge branch 'develop' into tdrr-clean-refactor

yardasol committed:
cleanup

yardasol committed:
fix scalar flux time derviatve scaling

jtramm committed:
Replace xtensor with internal Tensor/View classes (#3805)

Co-authored-by: John Tramm <jtramm@gmail.com>

yardasol committed:
update tests

yardasol committed:
Merge remote-tracking branch 'upstream/develop' into tdrr-clean-refactor

yardasol committed:
fast forward to use updated functions; add some missing flow control

yardasol committed:
fix dagmc unit test

yardasol committed:
remove cruft file

yardasol committed:
fix comment

yardasol committed:
model formatting fixes

yardasol committed:
Merge branch 'develop' into tdrr-clean-refactor

yardasol committed:
Revert "update tests"

This reverts commit 9c1a73f38348df266750070f937fa5de6b9fc03d.

yardasol committed:
Revert "fix scalar flux time derviatve scaling"

This reverts commit 774fb9b7302be4d8dfd6c9a1dca4fabcbf2db29e.
